repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -63,97 +76,67 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest)
-    if (removedHeaders.contains(wrappedHeaderName)) {
-      removedHeaders = removedHeaders.filterNot(_.equals(wrappedHeaderName))
+    val existingHeaders: List[String] = getHeadersScala(headerName) //this has to be done before we remove from the list,
+                                                                    // because getting this list is partially based on the contents of the removed list
+    if (removedHeaders.contains(headerName)) {
+      removedHeaders = removedHeaders.filterNot(_.equalsIgnoreCase(headerName))
     }
-    headerMap = headerMap + (wrappedHeaderName ->  (existingHeaders :+ headerValue))
+    headerMap = headerMap + (headerName -> (existingHeaders :+ headerValue))
   }
 
   override def addHeader(headerName: String, headerValue: String, quality: Double): Unit = addHeader(headerName, headerValue + ";q=" + quality)
 
   override def appendHeader(headerName: String, headerValue: String): Unit = {
-    val existingHeaders: List[String] = getHeadersScalaList(headerName)
+    val existingHeaders: List[String] = getHeadersScala(headerName)
     existingHeaders.headOption match {
-      case(Some(value)) => {
-        val newHeadValue :String = value + "," + headerValue
-        headerMap = headerMap + (HeaderName.wrap(headerName) -> (newHeadValue +: existingHeaders.tail))
-      }
-        case(None) => addHeader(headerName, headerValue)
+      case Some(value) =>
+        val newHeadValue: String = value + "," + headerValue
+        headerMap = headerMap + (headerName -> (newHeadValue +: existingHeaders.tail))
+      case None => addHeader(headerName, headerValue)
     }
   }
 
   override def appendHeader(headerName: String, headerValue: String, quality: Double): Unit = appendHeader(headerName, headerValue + ";q=" + quality)
 
   override def removeHeader(headerName: String): Unit = {
-    val wrappedHeaderName: HeaderName = HeaderName.wrap(headerName)
-    removedHeaders = removedHeaders + wrappedHeaderName
-    headerMap = headerMap.filterKeys(!_.equals(wrappedHeaderName))
+    removedHeaders = removedHeaders + headerName
+    headerMap = headerMap.filterKeys(!_.equalsIgnoreCase(headerName))
   }
 
-  case class HeaderWithParameters(value :String, parameters :Map[String, String])
-  case class HeaderWithQuailty(value :String, quality :Double)
-
-  def getValueWithQuality(headerValues :List[HeaderWithParameters]) :List[HeaderWithQuailty] = {
-    headerValues.map { header =>
-      HeaderWithQuailty(header.value, Option(header.parameters.getOrElse("q", "1.0")).map(_.toDouble).getOrElse(0.0))
+  def getPreferredHeader(headerName: String, getFun: String => List[String]): String = {
+    def parseQuality(headerValue: String): Double = {
+      //todo: what if the parameter value is not a number? Try and default to 1.0?
+      headerValue.split(";").tail.find(param => "q".equalsIgnoreCase(param.split("=")(0).trim))
```

Comment:
Alright. Making it 0 is easy enough to do using Try. Also, the assumption that we're looking at a non-splittable header is a fair point. Leave that to the user to decide which method to call. In which case I think this code still works as best it can.

As for clarity, I thought this was better than the three function and two case classes being used before. So. Much. Parsing. Logic. Still, you can write the parseQuality function however you'd like as long as it satisfies the signature, so if using substrings or something else is cleared, that's an easy change.